wasm/leb128/read.go

```diff
@@ -7,66 +7,81 @@
 package leb128
 
 import (
+	"errors"
 	"io"
 )
 
-// ReadVarUint32 reads a LEB128 encoded unsigned 32-bit integer from r, and
-// returns the integer value, and the error (if any).
-func ReadVarUint32(r io.Reader) (uint32, error) {
-	var (
-		b     = make([]byte, 1)
-		shift uint
-		res   uint32
-		err   error
-	)
-	for {
-		if _, err = io.ReadFull(r, b); err != nil {
-			return res, err
+// read uN defined in https://webassembly.github.io/spec/core/binary/values.html#binary-int
+// note: the result type is uint64, so N must <= 64.
+func readVarUint(r io.Reader, N uint) (uint64, error) {
+	p := make([]byte, 1)
+	_, err := r.Read(p)
+	if err != nil {
+		return 0, err
+	}
+	n := uint64(p[0])
+	if n < 1<<7 && n < 1<<N {
+		return n, nil
+	} else if n >= 1<<7 && N > 7 {
+		m, err := readVarUint(r, N-7)
+		if err != nil {
+			return 0, err
 		}
 
-		cur := uint32(b[0])
-		res |= (cur & 0x7f) << (shift)
-		if cur&0x80 == 0 {
-			return res, nil
+		return (1<<7)*m + (n - 1<<7), nil
+	}
+
+	return 0, errors.New("leb128: invalid uint")
+}
+
+// read sN defined in https://webassembly.github.io/spec/core/binary/values.html#binary-int
```

Comment:
```suggestion
// readVarint reads a signed integer of size n, defined in https://webassembly.github.io/spec/core/binary/values.html#binary-int
```

(same modification for the `note` below.)